File: day02.py
from pathlib import Path

shape_scores = {'rock': 1, 'paper': 2, 'scissors': 3}
result_scores = {'loss': 0, 'draw': 3, 'win': 6}
letter_shape_mapping = {
    'A': 'rock',
    'B': 'paper',
    'C': 'scissors',
    'X': 'rock',
    'Y': 'paper',
    'Z': 'scissors',
}
letter_result_mapping = {'X': 'loss', 'Y': 'draw', 'Z': 'win'}
required_shape = {
    ('rock', 'loss'): 'scissors',
    ('rock', 'draw'): 'rock',
    ('rock', 'win'): 'paper',
    ('scissors', 'loss'): 'paper',
    ('scissors', 'draw'): 'scissors',
    ('scissors', 'win'): 'rock',
    ('paper', 'loss'): 'rock',
    ('paper', 'draw'): 'paper',
    ('paper', 'win'): 'scissors',
}


def get_result_score(enemy_shape: str, my_shape: str) -> int:
    match (enemy_shape, my_shape):
        case m1, m2 if m1 == m2:
            return result_scores['draw']
        case ('rock', 'paper') | ('paper', 'scissors') | ('scissors', 'rock'):
            return result_scores['win']
        case ('rock', 'scissors') | ('paper', 'rock') | ('scissors', 'paper'):
            return result_scores['loss']
        case _:
            raise RuntimeError('Invalid case!')


def part1(rounds: list[(str, str)]) -> int:
    total_score = 0
    for enemy_shape_char, my_shape_char in rounds:
        enemy_shape = letter_shape_mapping[enemy_shape_char]
        my_shape = letter_shape_mapping[my_shape_char]
        total_score += shape_scores[my_shape] + get_result_score(enemy_shape, my_shape)
    return total_score


def part2(rounds: list[(str, str)]) -> int:
    total_score = 0
    for shape_char, result_char in rounds:
        enemy_shape = letter_shape_mapping[shape_char]
        result = letter_result_mapping[result_char]
        needed_shape = required_shape[(enemy_shape, result)]
        total_score += shape_scores[needed_shape] + result_scores[result]
    return total_score


if __name__ == '__main__':
    data = Path('inputs/day02.txt').read_text(encoding='utf-8')
    split_data = [line.split(" ") for line in data.splitlines()]
    answer1 = part1(split_data)
    print('Answer1:', answer1)  # 15572
    answer2 = part2(split_data)
    print('Answer2:', answer2)  # 16098

File: day02.kt
import java.io.File

enum class Shape {
    Rock, Paper, Scissors;

    fun getScore() = when (this) {
        Rock -> 1
        Paper -> 2
        Scissors -> 3
    }

    fun getInferior() = when (this) {
        Rock -> Scissors
        Paper -> Rock
        Scissors -> Paper
    }

    fun getSuperior() = when (this) {
        Rock -> Paper
        Paper -> Scissors
        Scissors -> Rock
    }

    companion object {
        fun fromChar(char: Char): Shape {
            return when (char) {
                'A' -> Rock
                'B' -> Paper
                'C' -> Scissors
                'X' -> Rock
                'Y' -> Paper
                'Z' -> Scissors
                else -> throw RuntimeException("Invalid argument: $char")
            }
        }
    }
}

data class Argument(val ch: Char)

enum class Outcome {
    Win, Draw, Loss;

    fun getScore() = when (this) {
        Loss -> 0
        Draw -> 3
        Win -> 6
    }

    companion object {
        fun fromArgument(arg: Argument) = when (arg.ch) {
            'X' -> Loss
            'Y' -> Draw
            'Z' -> Win
            else -> throw RuntimeException("invalid argument: ${arg.ch}")
        }
    }
}

data class Round(val enemyShape: Shape, val argument: Argument) {
    fun getOutcome(myShape: Shape): Outcome {
        return when (myShape) {
            enemyShape.getSuperior() -> Outcome.Win
            enemyShape.getInferior() -> Outcome.Loss
            else -> Outcome.Draw
        }
    }

    companion object {
        fun fromString(line: String): Round {
            val (shapeCh, argCh) = line.split(" ").map { it[0] }.take(2)
            return Round(Shape.fromChar(shapeCh), Argument(argCh))
        }
    }
}


fun preProcessData(lines: List<String>): List<Round> {
    return lines.map(Round::fromString)
}


fun part1(rounds: List<Round>): Int {
    var sum = 0
    for (round in rounds) {
        val myShape = Shape.fromChar(round.argument.ch)
        sum += myShape.getScore() + round.getOutcome(myShape).getScore()
    }
    return sum
}

fun part2(rounds: List<Round>): Int {
    var sum = 0
    for (round in rounds) {
        val outcome = Outcome.fromArgument(round.argument)
        val myShape = when (outcome) {
            Outcome.Win -> round.enemyShape.getSuperior()
            Outcome.Loss -> round.enemyShape.getInferior()
            Outcome.Draw -> round.enemyShape
        }
        sum += myShape.getScore() + outcome.getScore()
    }
    return sum
}


fun main() {
    val fileData = File("inputs/day02.txt").readLines()
    val processedData = preProcessData(fileData)
    val answer1 = part1(processedData)
    println("Answer 1: $answer1") // 15572
    val answer2 = part2(processedData)
    println("Answer 2: $answer2") // 16098
}